File: backend/src/modules/stock/stock.service.ts
import { Injectable, BadRequestException, NotFoundException } from '@nestjs/common'
import { InjectRepository } from '@nestjs/typeorm'
import { Repository, DataSource, EntityManager } from 'typeorm'
import { StockTransaction, TxType } from './entities/stock-transaction.entity'
import { Inventory } from '../inventory/inventory.entity'
import { StockValidator } from './domain/stock-validator'
import { TxnNoGenerator } from './domain/txn-no-generator'
import {
  CreateInboundDto, CreateOutboundDto,
  CreateAdjustmentDto, CreateMoveDto,
  CancelTxnDto, StockTxnQueryDto,
} from './dto/create-inbound.dto'

@Injectable()
export class StockService {
  constructor(
    @InjectRepository(StockTransaction) private txnRepo: Repository<StockTransaction>,
    @InjectRepository(Inventory) private invRepo: Repository<Inventory>,
    private dataSource: DataSource,
    private txnNoGen: TxnNoGenerator,
  ) {}

  // ────────────────────────────────────────
  // 입고
  // ────────────────────────────────────────
  async inbound(dto: CreateInboundDto, userId: string) {
    return this.dataSource.transaction(async (em) => {
      const inventory = await this.lockInventory(em, dto.productId, dto.locationId, dto.lotNumber)

      const beforeQty = inventory?.quantity ?? 0
      const afterQty  = beforeQty + dto.quantity
      StockValidator.assertNotNegative(afterQty, 'inbound')

      await this.upsertInventory(em, dto, afterQty, inventory)

      return this.saveTxn(em, {
        txType:      'INBOUND',
        qty:         dto.quantity,
        qtyBefore:   beforeQty,
        qtyAfter:    afterQty,
        productId:   dto.productId,
        locationId:  dto.locationId,
        warehouseId: dto.warehouseId,
        barcodeScanned: dto.barcodeUsed,
        lotNumber:   dto.lotNumber,
        expiryDate:  dto.expireDate,
        reason:      dto.reason,
        createdBy:   userId,
      })
    })
  }

  // ────────────────────────────────────────
  // 출고
  // ────────────────────────────────────────
  async outbound(dto: CreateOutboundDto, userId: string) {
    return this.dataSource.transaction(async (em) => {
      const inventory = await this.lockInventory(em, dto.productId, dto.locationId, dto.lotNumber)

      StockValidator.assertExists(inventory)
      StockValidator.assertSufficient(inventory, dto.quantity)

      const beforeQty = inventory.quantity
      const afterQty  = beforeQty - dto.quantity
      StockValidator.assertNotNegative(afterQty, 'outbound')

      await em.update(Inventory, inventory.id, {
        quantity: afterQty,
        version:  () => 'version + 1',
        lastSyncedAt: new Date(),
      })

      return this.saveTxn(em, {
        txType:      'OUTBOUND',
        qty:         -dto.quantity,
        qtyBefore:   beforeQty,
        qtyAfter:    afterQty,
        productId:   dto.productId,
        locationId:  dto.locationId,
        warehouseId: dto.warehouseId,
        barcodeScanned: dto.barcodeUsed,
        lotNumber:   dto.lotNumber,
        reason:      dto.reason,
        createdBy:   userId,
      })
    })
  }

  // ────────────────────────────────────────
  // 재고 조정
  // ────────────────────────────────────────
  async adjust(dto: CreateAdjustmentDto, userId: string) {
    return this.dataSource.transaction(async (em) => {
      const inventory = await this.lockInventory(em, dto.productId, dto.locationId, dto.lotNumber)

      const beforeQty = inventory?.quantity ?? 0
      const delta     = dto.adjustedQty - beforeQty
      StockValidator.assertNotNegative(dto.adjustedQty, 'adjustment')

      const txType: TxType = delta >= 0 ? 'ADJUST_INCREASE' : 'ADJUST_DECREASE'

      if (inventory) {
        await em.update(Inventory, inventory.id, {
          quantity: dto.adjustedQty,
          version:  () => 'version + 1',
          lastSyncedAt: new Date(),
        })
      } else {
        await em.insert(Inventory, {
          productId:   dto.productId,
          locationId:  dto.locationId,
          warehouseId: dto.warehouseId,
          quantity:    dto.adjustedQty,
          lotNumber:   dto.lotNumber ?? null,
        })
      }

      return this.saveTxn(em, {
        txType,
        qty:         delta,
        qtyBefore:   beforeQty,
        qtyAfter:    dto.adjustedQty,
        productId:   dto.productId,
        locationId:  dto.locationId,
        warehouseId: dto.warehouseId,
        reason:      dto.reason,
        createdBy:   userId,
      })
    })
  }

  // ────────────────────────────────────────
  // 위치 이동
  // ────────────────────────────────────────
  async move(dto: CreateMoveDto, userId: string) {
    if (dto.fromLocationId === dto.toLocationId) {
      throw new BadRequestException('출발지와 목적지가 동일합니다')
    }

    return this.dataSource.transaction(async (em) => {
      // 데드락 방지: location_id 오름차순으로 락 획득
      const [firstId, secondId] = [dto.fromLocationId, dto.toLocationId].sort()
      const [inv1, inv2] = await Promise.all([
        this.lockInventory(em, dto.productId, firstId,  dto.lotNumber),
        this.lockInventory(em, dto.productId, secondId, dto.lotNumber),
      ])

      const fromInv = firstId === dto.fromLocationId ? inv1 : inv2
      const toInv   = firstId === dto.toLocationId   ? inv1 : inv2

      StockValidator.assertExists(fromInv)
      StockValidator.assertSufficient(fromInv, dto.quantity)

      const fromBefore = fromInv.quantity
      const fromAfter  = fromBefore - dto.quantity
      const toBefore   = toInv?.quantity ?? 0
      const toAfter    = toBefore + dto.quantity

      await em.update(Inventory, fromInv.id, { quantity: fromAfter, version: () => 'version + 1' })

      if (toInv) {
        await em.update(Inventory, toInv.id, { quantity: toAfter, version: () => 'version + 1' })
      } else {
        await em.insert(Inventory, {
          productId:   dto.productId,
          locationId:  dto.toLocationId,
          warehouseId: dto.warehouseId,
          quantity:    dto.quantity,
          lotNumber:   dto.lotNumber ?? null,
        })
      }

      const moveOutTxn = await this.saveTxn(em, {
        txType:      'MOVE_OUT',
        qty:         -dto.quantity,
        qtyBefore:   fromBefore,
        qtyAfter:    fromAfter,
        productId:   dto.productId,
        locationId:  dto.fromLocationId,
        warehouseId: dto.warehouseId,
        reason:      dto.reason,
        createdBy:   userId,
      })

      await this.saveTxn(em, {
        txType:          'MOVE_IN',
        qty:             dto.quantity,
        qtyBefore:       toBefore,
        qtyAfter:        toAfter,
        productId:       dto.productId,
        locationId:      dto.toLocationId,
        warehouseId:     dto.warehouseId,
        reason:          dto.reason,
        referenceId:     moveOutTxn.id,
        referenceType:   'MOVE_OUT',
        createdBy:       userId,
      })

      return { moveOutTxnId: moveOutTxn.id, fromAfter, toAfter }
    })
  }

  // ────────────────────────────────────────
  // 입고 취소
  // ────────────────────────────────────────
  async cancelInbound(txnId: string, dto: CancelTxnDto, userId: string) {
    return this.dataSource.transaction(async (em) => {
      const original = await em.findOne(StockTransaction, {
        where: { id: txnId },
        lock:  { mode: 'pessimistic_write' },
      })
      if (!original) throw new NotFoundException('거래를 찾을 수 없습니다')
      StockValidator.assertNotCancelled(original.isCancelled)
      if (original.txType !== 'INBOUND') {
        throw new BadRequestException('입고 거래만 취소할 수 있습니다')
      }

      const inventory = await this.lockInventory(em, original.productId, original.locationId, original.lotNumber)
      StockValidator.assertExists(inventory)

      const cancelQty = original.qty  // 원거래 수량 (양수)
      if (inventory.quantity < cancelQty) {
        throw new BadRequestException({
          code:        'CANNOT_CANCEL',
          message:     '이미 출고된 수량이 있어 입고 취소가 불가합니다',
          currentQty:  inventory.quantity,
          cancelQty,
        })
      }

      const beforeQty = inventory.quantity
      const afterQty  = beforeQty - cancelQty
      StockValidator.assertNotNegative(afterQty, 'cancel inbound')

      await em.update(Inventory, inventory.id, { quantity: afterQty, version: () => 'version + 1' })
      await em.update(StockTransaction, txnId, {
        isCancelled: true,
        cancelledBy: userId,
        cancelledAt: new Date(),
      })

      return this.saveTxn(em, {
        txType:        'INBOUND_CANCEL',
        qty:           -cancelQty,
        qtyBefore:     beforeQty,
        qtyAfter:      afterQty,
        productId:     original.productId,
        locationId:    original.locationId,
        warehouseId:   original.warehouseId,
        referenceId:   txnId,
        referenceType: 'INBOUND',
        reason:        dto.reason,
        createdBy:     userId,
      })
    })
  }

  // ────────────────────────────────────────
  // 출고 취소
  // ────────────────────────────────────────
  async cancelOutbound(txnId: string, dto: CancelTxnDto, userId: string) {
    return this.dataSource.transaction(async (em) => {
      const original = await em.findOne(StockTransaction, {
        where: { id: txnId },
        lock:  { mode: 'pessimistic_write' },
      })
      if (!original) throw new NotFoundException('거래를 찾을 수 없습니다')
      StockValidator.assertNotCancelled(original.isCancelled)
      if (original.txType !== 'OUTBOUND') {
        throw new BadRequestException('출고 거래만 취소할 수 있습니다')
      }

      const inventory = await this.lockInventory(em, original.productId, original.locationId, original.lotNumber)

      const cancelQty = Math.abs(original.qty)
      const beforeQty = inventory?.quantity ?? 0
      const afterQty  = beforeQty + cancelQty

      await this.upsertInventory(em, {
        productId:   original.productId,
        locationId:  original.locationId,
        warehouseId: original.warehouseId,
        quantity:    afterQty,
        lotNumber:   original.lotNumber,
      }, afterQty, inventory)

      await em.update(StockTransaction, txnId, {
        isCancelled: true,
        cancelledBy: userId,
        cancelledAt: new Date(),
      })

      return this.saveTxn(em, {
        txType:        'OUTBOUND_CANCEL',
        qty:           cancelQty,
        qtyBefore:     beforeQty,
        qtyAfter:      afterQty,
        productId:     original.productId,
        locationId:    original.locationId,
        warehouseId:   original.warehouseId,
        referenceId:   txnId,
        referenceType: 'OUTBOUND',
        reason:        dto.reason,
        createdBy:     userId,
      })
    })
  }

  // ────────────────────────────────────────
  // 거래 내역 조회
  // ────────────────────────────────────────
  async findTransactions(q: StockTxnQueryDto) {
    const { productId, locationId, warehouseId, txType, from, to, performedBy, page = 1, limit = 100 } = q

    const qb = this.txnRepo
      .createQueryBuilder('t')
      .leftJoinAndSelect('t.product', 'p')
      .leftJoinAndSelect('t.location', 'loc')
      .leftJoinAndSelect('t.user', 'u')
      .orderBy('t.createdAt', 'DESC')

    if (productId)   qb.andWhere('t.productId = :productId', { productId })
    if (locationId)  qb.andWhere('t.locationId = :locationId', { locationId })
    if (warehouseId) qb.andWhere('t.warehouseId = :warehouseId', { warehouseId })
    if (txType)      qb.andWhere('t.txType = :txType', { txType })
    if (performedBy) qb.andWhere('t.createdBy = :performedBy', { performedBy })
    if (from)        qb.andWhere('t.createdAt >= :from', { from: new Date(from) })
    if (to)          qb.andWhere('t.createdAt <= :to', { to: new Date(to + 'T23:59:59Z') })

    const [items, total] = await qb.take(limit).skip((page - 1) * limit).getManyAndCount()
    return { items, total, page, limit, totalPages: Math.ceil(total / limit) }
  }

  // ────────────────────────────────────────
  // Private helpers
  // ────────────────────────────────────────
  private async lockInventory(
    em: EntityManager,
    productId: string,
    locationId: string,
    lotNumber?: string,
  ): Promise<Inventory | null> {
    const qb = em
      .createQueryBuilder(Inventory, 'inv')
      .setLock('pessimistic_write')
      .where('inv.product_id = :productId AND inv.location_id = :locationId', { productId, locationId })

    if (lotNumber) {
      qb.andWhere('inv.lot_number = :lotNumber', { lotNumber })
    } else {
      qb.andWhere('inv.lot_number IS NULL')
    }

    return qb.getOne()
  }

  private async upsertInventory(
    em: EntityManager,
    dto: { productId: string; locationId: string; warehouseId: string; lotNumber?: string; expireDate?: string; quantity?: number },
    newQty: number,
    existing: Inventory | null,
  ) {
    if (existing) {
      await em.update(Inventory, existing.id, {
        quantity:    newQty,
        version:     () => 'version + 1',
        lastSyncedAt: new Date(),
      })
    } else {
      await em.insert(Inventory, {
        productId:   dto.productId,
        locationId:  dto.locationId,
        warehouseId: dto.warehouseId,
        quantity:    newQty,
        lotNumber:   dto.lotNumber ?? null,
        expireDate:  (dto as any).expireDate ?? null,
      })
    }
  }

  private async saveTxn(em: EntityManager, data: Partial<StockTransaction> & { txType: TxType }): Promise<StockTransaction> {
    const txnNo = await this.txnNoGen.generate('TXN')
    StockValidator.assertIntegrity(data.qtyBefore!, data.qty!, data.qtyAfter!)
    return em.save(StockTransaction, { ...data, txnNo } as StockTransaction)
  }
}

File: backend/src/main/java/com/wmspro/domain/stock/StockService.java
package com.wmspro.domain.stock;

import com.wmspro.common.exception.BusinessException;
import com.wmspro.common.exception.ErrorCode;
import com.wmspro.domain.inventory.Inventory;
import com.wmspro.domain.inventory.InventoryRepository;
import com.wmspro.domain.stock.dto.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class StockService {

    private final StockTransactionRepository txnRepo;
    private final InventoryRepository        invRepo;
    private final JdbcTemplate               jdbcTemplate;

    // ────────────────────────────────────────────────────────────
    // 입고
    // ────────────────────────────────────────────────────────────
    @Transactional
    public StockTransaction inbound(InboundRequest req, UUID userId) {
        Optional<Inventory> invOpt = lockInventory(req.productId, req.locationId, req.lotNumber);

        int beforeQty = invOpt.map(Inventory::getQuantity).orElse(0);
        int afterQty  = beforeQty + req.quantity;
        assertNotNegative(afterQty, "inbound");

        if (invOpt.isPresent()) {
            Inventory inv = invOpt.get();
            inv.setQuantity(afterQty);
            inv.setLastSyncedAt(Instant.now());
            invRepo.save(inv);
        } else {
            invRepo.save(Inventory.builder()
                .productId(req.productId)
                .locationId(req.locationId)
                .warehouseId(req.warehouseId)
                .quantity(req.quantity)
                .lotNumber(req.lotNumber)
                .expireDate(req.expireDate)
                .lastSyncedAt(Instant.now())
                .build());
        }

        return saveTxn(StockTransaction.builder()
            .txType(TxType.INBOUND)
            .qty(req.quantity)
            .qtyBefore(beforeQty)
            .qtyAfter(afterQty)
            .productId(req.productId)
            .locationId(req.locationId)
            .warehouseId(req.warehouseId)
            .barcodeScanned(req.barcodeUsed)
            .lotNumber(req.lotNumber)
            .expiryDate(req.expireDate)
            .reason(req.reason)
            .memo(req.memo)
            .createdBy(userId)
            .build());
    }

    // ────────────────────────────────────────────────────────────
    // 출고 — 비관적 락 필수
    // ────────────────────────────────────────────────────────────
    @Transactional
    public StockTransaction outbound(OutboundRequest req, UUID userId) {
        Inventory inv = lockInventory(req.productId, req.locationId, req.lotNumber)
            .orElseThrow(() -> new BusinessException(ErrorCode.INVENTORY_NOT_FOUND));

        assertSufficient(inv, req.quantity);

        int beforeQty = inv.getQuantity();
        int afterQty  = beforeQty - req.quantity;
        assertNotNegative(afterQty, "outbound");

        inv.setQuantity(afterQty);
        inv.setLastSyncedAt(Instant.now());
        invRepo.save(inv);

        return saveTxn(StockTransaction.builder()
            .txType(TxType.OUTBOUND)
            .qty(-req.quantity)
            .qtyBefore(beforeQty)
            .qtyAfter(afterQty)
            .productId(req.productId)
            .locationId(req.locationId)
            .warehouseId(req.warehouseId)
            .barcodeScanned(req.barcodeUsed)
            .lotNumber(req.lotNumber)
            .reason(req.reason)
            .memo(req.memo)
            .createdBy(userId)
            .build());
    }

    // ────────────────────────────────────────────────────────────
    // 재고 조정
    // ────────────────────────────────────────────────────────────
    @Transactional
    public StockTransaction adjust(AdjustRequest req, UUID userId) {
        assertNotNegative(req.adjustedQty, "adjustment");

        Optional<Inventory> invOpt = lockInventory(req.productId, req.locationId, req.lotNumber);
        int beforeQty = invOpt.map(Inventory::getQuantity).orElse(0);
        int delta     = req.adjustedQty - beforeQty;

        if (invOpt.isPresent()) {
            Inventory inv = invOpt.get();
            inv.setQuantity(req.adjustedQty);
            inv.setLastSyncedAt(Instant.now());
            invRepo.save(inv);
        } else {
            invRepo.save(Inventory.builder()
                .productId(req.productId)
                .locationId(req.locationId)
                .warehouseId(req.warehouseId)
                .quantity(req.adjustedQty)
                .lotNumber(req.lotNumber)
                .lastSyncedAt(Instant.now())
                .build());
        }

        TxType txType = delta >= 0 ? TxType.ADJUST_INCREASE : TxType.ADJUST_DECREASE;

        return saveTxn(StockTransaction.builder()
            .txType(txType)
            .qty(delta)
            .qtyBefore(beforeQty)
            .qtyAfter(req.adjustedQty)
            .productId(req.productId)
            .locationId(req.locationId)
            .warehouseId(req.warehouseId)
            .lotNumber(req.lotNumber)
            .reason(req.reason)
            .memo(req.memo)
            .createdBy(userId)
            .build());
    }

    // ────────────────────────────────────────────────────────────
    // 위치 이동 — 데드락 방지: locationId 오름차순으로 락 획득
    // ────────────────────────────────────────────────────────────
    @Transactional
    public void move(MoveRequest req, UUID userId) {
        if (req.fromLocationId.equals(req.toLocationId)) {
            throw new BusinessException(ErrorCode.SAME_LOCATION);
        }

        // 항상 정렬된 순서로 락 획득
        List<UUID> sortedIds = List.of(req.fromLocationId, req.toLocationId)
            .stream().sorted().toList();

        List<Inventory> locked = invRepo.findWithLockMultiple(req.productId, sortedIds);

        Inventory fromInv = locked.stream()
            .filter(i -> i.getLocationId().equals(req.fromLocationId))
            .findFirst()
            .orElseThrow(() -> new BusinessException(ErrorCode.INVENTORY_NOT_FOUND));

        Inventory toInv = locked.stream()
            .filter(i -> i.getLocationId().equals(req.toLocationId))
            .findFirst()
            .orElse(null);

        assertSufficient(fromInv, req.quantity);

        int fromBefore = fromInv.getQuantity();
        int fromAfter  = fromBefore - req.quantity;
        int toBefore   = toInv != null ? toInv.getQuantity() : 0;
        int toAfter    = toBefore + req.quantity;

        fromInv.setQuantity(fromAfter);
        fromInv.setLastSyncedAt(Instant.now());
        invRepo.save(fromInv);

        if (toInv != null) {
            toInv.setQuantity(toAfter);
            toInv.setLastSyncedAt(Instant.now());
            invRepo.save(toInv);
        } else {
            invRepo.save(Inventory.builder()
                .productId(req.productId)
                .locationId(req.toLocationId)
                .warehouseId(req.warehouseId)
                .quantity(req.quantity)
                .lotNumber(req.lotNumber)
                .lastSyncedAt(Instant.now())
                .build());
        }

        StockTransaction moveOut = saveTxn(StockTransaction.builder()
            .txType(TxType.MOVE_OUT)
            .qty(-req.quantity)
            .qtyBefore(fromBefore)
            .qtyAfter(fromAfter)
            .productId(req.productId)
            .locationId(req.fromLocationId)
            .warehouseId(req.warehouseId)
            .reason(req.reason)
            .createdBy(userId)
            .build());

        saveTxn(StockTransaction.builder()
            .txType(TxType.MOVE_IN)
            .qty(req.quantity)
            .qtyBefore(toBefore)
            .qtyAfter(toAfter)
            .productId(req.productId)
            .locationId(req.toLocationId)
            .warehouseId(req.warehouseId)
            .referenceId(moveOut.getId())
            .referenceType("MOVE_OUT")
            .reason(req.reason)
            .createdBy(userId)
            .build());
    }

    // ────────────────────────────────────────────────────────────
    // 입고 취소
    // ────────────────────────────────────────────────────────────
    @Transactional
    public StockTransaction cancelInbound(UUID txnId, String reason, UUID userId) {
        StockTransaction original = txnRepo.findWithLock(txnId)
            .orElseThrow(() -> new BusinessException(ErrorCode.TRANSACTION_NOT_FOUND));

        if (original.isCancelled()) throw new BusinessException(ErrorCode.ALREADY_CANCELLED);
        if (original.getTxType() != TxType.INBOUND) throw new BusinessException(ErrorCode.INVALID_TX_TYPE);

        Inventory inv = lockInventory(original.getProductId(), original.getLocationId(), original.getLotNumber())
            .orElseThrow(() -> new BusinessException(ErrorCode.INVENTORY_NOT_FOUND));

        if (inv.getQuantity() < original.getQty()) {
            throw new BusinessException(ErrorCode.CANNOT_CANCEL);
        }

        int beforeQty = inv.getQuantity();
        int afterQty  = beforeQty - original.getQty();
        assertNotNegative(afterQty, "cancel inbound");

        inv.setQuantity(afterQty);
        invRepo.save(inv);

        original.setCancelled(true);
        original.setCancelledBy(userId);
        original.setCancelledAt(Instant.now());
        txnRepo.save(original);

        return saveTxn(StockTransaction.builder()
            .txType(TxType.INBOUND_CANCEL)
            .qty(-original.getQty())
            .qtyBefore(beforeQty)
            .qtyAfter(afterQty)
            .productId(original.getProductId())
            .locationId(original.getLocationId())
            .warehouseId(original.getWarehouseId())
            .referenceId(txnId)
            .referenceType("INBOUND")
            .reason(reason)
            .createdBy(userId)
            .build());
    }

    // ────────────────────────────────────────────────────────────
    // 출고 취소
    // ────────────────────────────────────────────────────────────
    @Transactional
    public StockTransaction cancelOutbound(UUID txnId, String reason, UUID userId) {
        StockTransaction original = txnRepo.findWithLock(txnId)
            .orElseThrow(() -> new BusinessException(ErrorCode.TRANSACTION_NOT_FOUND));

        if (original.isCancelled()) throw new BusinessException(ErrorCode.ALREADY_CANCELLED);
        if (original.getTxType() != TxType.OUTBOUND) throw new BusinessException(ErrorCode.INVALID_TX_TYPE);

        Optional<Inventory> invOpt = lockInventory(original.getProductId(), original.getLocationId(), original.getLotNumber());

        int cancelQty = Math.abs(original.getQty());
        int beforeQty = invOpt.map(Inventory::getQuantity).orElse(0);
        int afterQty  = beforeQty + cancelQty;

        if (invOpt.isPresent()) {
            Inventory inv = invOpt.get();
            inv.setQuantity(afterQty);
            inv.setLastSyncedAt(Instant.now());
            invRepo.save(inv);
        } else {
            invRepo.save(Inventory.builder()
                .productId(original.getProductId())
                .locationId(original.getLocationId())
                .warehouseId(original.getWarehouseId())
                .quantity(cancelQty)
                .lotNumber(original.getLotNumber())
                .lastSyncedAt(Instant.now())
                .build());
        }

        original.setCancelled(true);
        original.setCancelledBy(userId);
        original.setCancelledAt(Instant.now());
        txnRepo.save(original);

        return saveTxn(StockTransaction.builder()
            .txType(TxType.OUTBOUND_CANCEL)
            .qty(cancelQty)
            .qtyBefore(beforeQty)
            .qtyAfter(afterQty)
            .productId(original.getProductId())
            .locationId(original.getLocationId())
            .warehouseId(original.getWarehouseId())
            .referenceId(txnId)
            .referenceType("OUTBOUND")
            .reason(reason)
            .createdBy(userId)
            .build());
    }

    // ────────────────────────────────────────────────────────────
    // Private helpers
    // ────────────────────────────────────────────────────────────
    private Optional<Inventory> lockInventory(UUID productId, UUID locationId, String lotNumber) {
        if (lotNumber != null && !lotNumber.isBlank()) {
            return invRepo.findWithLockAndLot(productId, locationId, lotNumber);
        }
        return invRepo.findWithLock(productId, locationId);
    }

    private void assertSufficient(Inventory inv, int requested) {
        int available = inv.getAvailableQty();
        if (available < requested) {
            throw new BusinessException(ErrorCode.INSUFFICIENT_STOCK,
                java.util.Map.of(
                    "available", available,
                    "requested", requested,
                    "locationId", inv.getLocationId()
                ));
        }
    }

    private void assertNotNegative(int qty, String context) {
        if (qty < 0) throw new BusinessException(ErrorCode.STOCK_NEGATIVE);
    }

    private StockTransaction saveTxn(StockTransaction txn) {
        // 정합성 검증: before + delta == after
        if (txn.getQtyBefore() + txn.getQty() != txn.getQtyAfter()) {
            log.error("재고 정합성 오류: before={}, qty={}, after={}",
                txn.getQtyBefore(), txn.getQty(), txn.getQtyAfter());
            throw new BusinessException(ErrorCode.STOCK_INTEGRITY_ERROR);
        }
        txn.setTxnNo(generateTxnNo());
        return txnRepo.save(txn);
    }

    private String generateTxnNo() {
        Long seq = jdbcTemplate.queryForObject("SELECT nextval('txn_seq')", Long.class);
        String date = LocalDate.now(ZoneId.of("Asia/Seoul"))
            .format(DateTimeFormatter.ofPattern("yyyyMMdd"));
        return String.format("TXN-%s-%06d", date, seq);
    }
}
